### PR TITLE
Synchronize and cleanup instruction processor lists

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Multiple places exist that create genesis blocks and each maintains its own list.

#### Summary of Changes

Use a single default list for instruction processors.  Individual tests or use cases can still override the default.

Fixes #

### CODE REVIEW

**Comment by jackcmay** [comment]
@mvines `ci/order-crates-for-publishing.py` is failing.  Any suggestions on how to address these potential circular dependencies?
```
Error: Circular dependency suspected between these packages: solana-validator solana-exchange-api solana-config-program solana-genesis solana-bench-streamer solana-runtime solana-replicator solana-bpf-programs solana-validator-info solana-programs solana-failure-program solana solana-bench-tps solana-ledger-tool solana-config-api solana-exchange-program solana-wallet solana-budget-program solana-install solana-budget-api solana-stake-tests solana-bench-exchange solana-librapay-api solana-gossip solana-storage-program
```

**Comment by mvines** [comment]
Is there an actual circular dependency or is the script just wrong?

**Comment by jackcmay** [comment]
@mvines 

Crate-wise I think its real.  For example, `runtime -> programs -> exchange_api -> runtime`

But I'm kinda surprised we haven't faced this earlier, or maybe we did and we got around it by shoving some of the program's id into runtime...  Either way this will probably come up again and I'm not familiar with the publishing rules to know what the problem really is or how to maybe address it.  Cargo seems to take care of the circular references fine, I guess publish is different?

**Comment by mvines** [comment]
> maybe we did and we got around it by shoving some of the program's id into runtime

Yeah that or in sdk/

Yeah publish cannot have a circular dependency since you can’t publish multiple crates atomically

File: runtime/Cargo.toml [hunk 1]
```diff
@@ -30,7 +30,7 @@ solana-measure = { path = "../measure", version = "0.18.0-pre0" }
 solana-metrics = { path = "../metrics", version = "0.18.0-pre0" }
 solana-bpf-loader-api = { path = "../programs/bpf_loader_api", version = "0.18.0-pre0" }
 solana-bpf-loader-program = { path = "../programs/bpf_loader_program", version = "0.18.0-pre0" }
-solana-noop-program = { path = "../programs/noop_program", version = "0.18.0-pre0" }
+solana-programs = { path = "../programs", version = "0.18.0-pre0" }
```
Comment (mvines) [inline runtime/Cargo.toml]: This dep looks backwards.  Runtime can’t generically depend on programs

File: programs/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+[package]
+name = "solana-programs"
```
Comment (mvines) [inline programs/Cargo.toml]: call this “solana-core-programs” perhaps, and move it into “core-programs/“

File: runtime/Cargo.toml (hunk 1 above)
Comment (mvines) [reply runtime/Cargo.toml]: I think you need two lists: (1) the minimal set that the runtime needs to function, “runtime-programs/“, and (2) everybody else, “core-programs”

File: programs/Cargo.toml (hunk 2 above)
Comment (jackcmay) [reply programs/Cargo.toml]: Don't think that would resolve dependency bit i think i can remove runtime's dep on programs, will give that a shot

Comment (mvines) [reply programs/Cargo.toml]: Yeah that comment was just about cleaning up the name.  It’s weird to just call it “solana-programs” imo, that’s too general of a name

Comment (jackcmay) [reply programs/Cargo.toml]: This crate is used by more than just core, What do you think about "default-programs" at "/default-programs/".

Comment (mvines) [reply programs/Cargo.toml]: Are all the dependencies (aside from runtime/) downstream of `core/`?  if so maybe this turns into `core/src/default_programs.rs`?   I'm not seeing a lot of value having this as a separate crate right now

Comment (jackcmay) [reply programs/Cargo.toml]: /genesis also depends on it

Comment (mvines) [reply programs/Cargo.toml]: 🤔 I don't like "default-programs" cause it means different things to different parts of the tree.  The default programs in runtime/ are not the same as the default programs in genesis/ (or local_cluster).

This naming convention like this clearly positions the programs next to the crate that primarily depends on them:
* `runtime-programs/`
* `core-programs/` (yeah genesis/ depends on core-programs/ too, but the "stronger" dependency is from core/)

Also an instruction_processor is not a thing, ITYM program.  Until we settle on the final name for program, using two names for the one thing just makes our world more confusing!

File: runtime/src/genesis_utils.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+//use solana_programs::get_default_native_instruction_processors;
```
Comment (mvines) [inline runtime/src/genesis_utils.rs]: nit: remove

File: runtime/src/genesis_utils.rs [hunk 4]
```diff
@@ -66,6 +67,7 @@ pub fn create_genesis_block_with_leader(
                 ),
             ),
         ])
+        // Bare minimum
```
Comment (mvines) [inline runtime/src/genesis_utils.rs]: nit: use more words.  "Bare minimum ..... what"?

**Comment by mvines** [comment]
💔 what about my nits

File: runtime/src/genesis_utils.rs (hunk 3 above)
Comment (sakridge) [reply runtime/src/genesis_utils.rs]: removed in https://github.com/solana-labs/solana/pull/5333

File: runtime/src/genesis_utils.rs (hunk 4 above)
Comment (sakridge) [reply runtime/src/genesis_utils.rs]: sentence finished in https://github.com/solana-labs/solana/pull/5333

**Pull request merged**